File: solutions/c/space-age/1/space_age.c
#include "space_age.h"

float age(planet_t planet, int64_t seconds) {
  const float earth_age = seconds / 365.25 / 24 / 60 / 60;

  switch (planet) {
  case EARTH:
    return earth_age;
  case MERCURY:
    return earth_age / 0.2408467;
  case VENUS:
    return earth_age / 0.61519726;
  case MARS:
    return earth_age / 1.8808158;
  case JUPITER:
    return earth_age / 11.862615;
  case SATURN:
    return earth_age / 29.447498;
  case URANUS:
    return earth_age / 84.016846;
  case NEPTUNE:
    return earth_age / 164.79132;
  default:
    return INVALID_PLANET;
  }
}

File: solutions/haskell/space-age/1/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

roundTo n x = (fromInteger $ round $ x * (10^n)) / (10.0^^n)

ageOn :: Planet -> Float -> Float
ageOn planet seconds = roundTo 2 (seconds / year)
  where
    year = case planet of
      Earth -> 31557600
      Mercury -> 31557600 * 0.2408467
      Venus -> 31557600 * 0.61519726
      Mars -> 31557600 * 1.8808158
      Jupiter -> 31557600 * 11.862615
      Saturn -> 31557600 * 29.447498
      Uranus -> 31557600 * 84.016846
      Neptune -> 31557600 * 164.79132